### PR TITLE
Time travel across chains

### PR DESCRIPTION
**Pull request by bipll**
This commit introduces methods `MainChain::TimeTravel` and `MainChainProtocol::TimeTravel`. Unlike `Get\i*Chain`, they allow to travel in both directions, up- and downchain.

### CODE REVIEW

File: libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp]: I would keep the enums the same for the moment, shouldn't really be an issue but we should get in the habit of trying to keep backward compat.

File: libs/ledger/include/ledger/chain/main_chain.hpp
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/chain/main_chain.hpp]: Is this `std::` prefix necessary. We haven't done it in the rest of the code base. Happy to change it if the team feels the same way but this is something that we should do in one operation

File: libs/ledger/src/chain/main_chain.cpp
Comment (ejfitzgerald) [inline libs/ledger/src/chain/main_chain.cpp:410]: :+1:

File: scripts/apply_style.py
Comment (ejfitzgerald) [inline scripts/apply_style.py]: @wkaluza FYI

File: libs/meta/include/meta/type_traits.hpp
Comment (ejfitzgerald) [inline libs/meta/include/meta/type_traits.hpp]: These don't appear to be used in conjunction with this PR. Can we separate them into there own PR if that is the case

File: libs/meta/include/meta/value_util.hpp
Comment (ejfitzgerald) [inline libs/meta/include/meta/value_util.hpp]: Same for this

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]

File: libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp
Comment (bipll) [reply libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp]: It's just the numbering felt a big confusing to me. If subchain is 3 then what should time travel be? Is it 2? Or 5? And where did the original gap come from?

Comment (ejfitzgerald) [reply libs/ledger/include/ledger/protocols/main_chain_rpc_protocol.hpp]: Oh yep it is confusing. There used to be another RPC method that was removed can't remember which one it was but it was a long time ago I believe. 

It's not a big thing, but in the future these will need to stabilise. So I think you should use `2` or `4` as a value. I am happy with either. 😁

File: libs/ledger/include/ledger/chain/main_chain.hpp
Comment (bipll) [reply libs/ledger/include/ledger/chain/main_chain.hpp]: Afaik, cstdint defines them in std. I wasn't able to find anything about them being injected in top-level so, as I understand, while everyone does it's still non-standard. And a separate PR would really be huge and cover almost the whole codebase. >_< No idea how it should be handled the most graceful.

Comment (bipll) [reply libs/ledger/include/ledger/chain/main_chain.hpp]: I have added `using`s to main_chain.hpp as a preliminary solution, IMO this needs to be done at the rootest include of this whole codebase but not in this PR.

File: scripts/apply_style.py
Comment (wkaluza) [reply scripts/apply_style.py]: I'd rather these changes were reverted, or at least tied more neatly into existing code. If the intent was to restrict the reformatting to a filesystem subtree, it's as simple as plumbing in the cwd into the `walk_source_directories` generator function instead of the `PROJECT_ROOT` constant. The rest of these changes are unnecessary. How many files with forbidden read or write access do we expect in a local working copy? I'm also not happy with introducing an execution path where `commit` is hijacked to be a list of files instead of a string specifying a Git ref. Apart from the fact that this is not what `commit` is for, this kind of type ambiguities and reliance on `isinstance` makes the code hard to follow and maintain. Just because Python allows it doesn't mean it's a good idea.

**Review by wkaluza — CHANGES_REQUESTED** [review]
See comment under apply_style.py

File: scripts/apply_style.py
Comment (bipll) [reply scripts/apply_style.py]: @wkaluza As I remember, these changes turned out necessary as otherwise, at some moment of time, apply_style.py wasn't able to find any file at all to process, unsure how it happened. Btw, I've added (slightly incomplete) path support in the command line but then it somehow disappeared from the script again few months ago.

Comment (wkaluza) [reply scripts/apply_style.py]: Seeing that no one else reported issues, this appears to be a problem local to your machine, and that's where I would seek to resolve it. In any case, these changes have nothing to do with this PR, so rather than clutter up this discussion, please revert.

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: libs/ledger/include/ledger/chain/main_chain.hpp
Comment (ejfitzgerald) [inline libs/ledger/include/ledger/chain/main_chain.hpp]: It might be non-standard but the rest of the code base doesn't using the std prefix. Lets ditch this and then we can cross the bridge when it doesn't work on a new platform in the future

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Looks like a good starting position. I would remove the type_utils stuff for the moment because they don't look like they are used in this change

**Review by wkaluza — DISMISSED** [review]
@ejfitzgerald now that `apply_style.py` is fixed, I'm withdrawing my request for changes and leave the rest in your hands.

**Comment by ejfitzgerald** [comment]
@bipll not sure if you are waiting on my review. From my side I think there are the following changes for this PR:

- [ ] Revert meta template changes (this should be a new PR)
- [ ] Remove the `using std::uint32_t` in the main chain header

Other than that this looks good. So lets make these changes and we can land this feature

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**
